sources/api/corndog/src/main.rs

```diff
@@ -104,23 +109,104 @@ where
+/// Validates sysctl key/value pairs for systemd-sysctl compatibility
+fn validate_sysctl_entry<K>(key: K, value: &str) -> Result<()> 
 where
     K: AsRef<str>,
 {
+    let key = key.as_ref();
+    
+    // Basic validation of key format
+    if !key.chars().all(|c| c.is_ascii_alphanumeric() || c == '.' || c == '_' || c == '-' || c == '*') {
+        return error::InvalidSysctlKeySnafu { key: key.to_string() }.fail();
+    }
+
+    // Ensure value doesn't contain newlines which could break the config file format
+    if value.contains('\n') {
+        return error::InvalidSysctlValueSnafu { 
+            key: key.to_string(),
+            value: value.to_string() 
+        }.fail();
+    }
+
+    Ok(())
+}
+
+/// Creates a backup of the existing sysctl config file
+fn backup_sysctl_config() -> Result<Option<NamedTempFile>> {
+    const SYSCTL_CONFIG: &str = "/etc/sysctl.d/95-corndog.conf";
+    
+    if !Path::new(SYSCTL_CONFIG).exists() {
+        return Ok(None);
+    }
+
+    let mut temp_file = NamedTempFile::new().context(error::TempFileSnafu)?;
+    let contents = fs::read_to_string(SYSCTL_CONFIG).context(error::ReadSysctlConfigSnafu)?;
+    temp_file.write_all(contents.as_bytes()).context(error::WriteTempFileSnafu)?;
+    
+    Ok(Some(temp_file))
+}
+
+/// Restores sysctl config from backup
+fn restore_sysctl_config(backup: NamedTempFile) -> Result<()> {
+    const SYSCTL_CONFIG: &str = "/etc/sysctl.d/95-corndog.conf";
+    
+    fs::copy(backup.path(), SYSCTL_CONFIG).context(error::RestoreBackupSnafu)?;
+    Ok(())
+}
+
+/// Restarts systemd-sysctl service
+fn restart_sysctl() -> Result<()> {
+    let status = Command::new("systemctl")
+        .args(["restart", "systemd-sysctl"])
+        .status()
+        .context(error::SystemctlSnafu)?;
+
+    if !status.success() {
+        return error::SystemctlFailedSnafu.fail();
+    }
+    
+    Ok(())
+}
+
+/// Applies the requested sysctls to the system using systemd-sysctl.
+fn set_sysctls<K>(sysctls: HashMap<K, String>) -> Result<()>
+where
+    K: AsRef<str>,
+{
+    // TODO: make the priority configurable?
+    const SYSCTL_CONFIG: &str = "/etc/sysctl.d/95-corndog.conf";
+
+    // Validate all entries first
+    for (key, value) in &sysctls {
+        validate_sysctl_entry(key, value)?;
+    }
+
+    // Create backup of existing config
+    let backup = backup_sysctl_config()?;
+
+    // Write new config
+    let mut config = String::new();
     for (key, value) in sysctls {
-        let key = key.as_ref();
-        let path = sysctl_path(key);
-        if let Err(e) = fs::write(path, value) {
-            // We don't fail because sysctl keys can vary between kernel versions and depend on
-            // loaded modules.  It wouldn't be possible to deploy settings to a mixed-kernel fleet
-            // if newer sysctl values failed on your older kernels, for example, and we believe
-            // it's too cumbersome to have to specify in settings which keys are allowed to fail.
```

Comment:
The current implementation of this PR doesn't maintain the functionality mentioned in this comment, but we could pretty easily achieve this by prepending `-` to all the sysctls passed in by default (which would then remove the need for the backup-modify-restore shuffle, which might make things simpler overall). Open to feedback/direction here.